File: pkg/console/model/service.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package model

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/duke-git/lancet/v2/strutil"
	"github.com/gin-gonic/gin"

	"github.com/apache/dubbo-admin/pkg/common/constants"
	coremodel "github.com/apache/dubbo-admin/pkg/core/resource/model"
)

type ServiceSearchReq struct {
	coremodel.PageReq

	ServiceName string `form:"serviceName" json:"serviceName"`
	Keywords    string `form:"keywords" json:"keywords"`
	Mesh        string `form:"mesh" json:"mesh"`
}

func NewServiceSearchReq() *ServiceSearchReq {
	return &ServiceSearchReq{
		PageReq: coremodel.PageReq{
			PageOffset: 0,
			PageSize:   15,
		},
	}
}

type ServiceSearchResp struct {
	ServiceName     string `json:"serviceName"`
	Version         string `json:"version"`
	Group           string `json:"group"`
	ProviderAppName string `json:"providerAppName,omitempty"`
	ConsumerAppName string `json:"consumerAppName,omitempty"`
}

type ByServiceName []*ServiceSearchResp

func (a ByServiceName) Len() int { return len(a) }

func (a ByServiceName) Less(i, j int) bool {
	return a[i].ServiceName < a[j].ServiceName
}

func (a ByServiceName) Swap(i, j int) { a[i], a[j] = a[j], a[i] }

type ServiceTabDistributionReq struct {
	ServiceName     string `json:"serviceName"  form:"serviceName" binding:"required"`
	Version         string `json:"version"  form:"version"`
	Group           string `json:"group"  form:"group"`
	Side            string `json:"side" form:"side"  binding:"required"`
	Mesh            string `json:"mesh" form:"mesh" binding:"required"`
	ProviderAppName string `json:"providerAppName"  form:"providerAppName"`
	Keywords        string `json:"keywords"  form:"keywords"`
	coremodel.PageReq
}

type ServiceTabDistributionResp struct {
	AppName      string            `json:"appName"`
	InstanceName string            `json:"instanceName"`
	Endpoint     string            `json:"endpoint"`
	TimeOut      string            `json:"timeOut"`
	Retries      string            `json:"retries"`
	Params       map[string]string `json:"params"`
}

type ByServiceInstanceName []*ServiceTabDistributionResp

func (a ByServiceInstanceName) Len() int { return len(a) }

func (a ByServiceInstanceName) Less(i, j int) bool {
	return a[i].InstanceName < a[j].InstanceName
}

func (a ByServiceInstanceName) Swap(i, j int) { a[i], a[j] = a[j], a[i] }

type ServiceTabDistribution struct {
	AppName      string
	InstanceName string
	Endpoint     string
	TimeOut      string
	Retries      string
}

type BaseServiceReq struct {
	ServiceName string `json:"serviceName"`
	Group       string `json:"group"`
	Version     string `json:"version"`
	Mesh        string `json:"mesh"`
}

func (s *BaseServiceReq) Query(c *gin.Context) error {
	s.ServiceName = strings.TrimSpace(c.Query("serviceName"))
	if strutil.IsBlank(s.ServiceName) {
		return fmt.Errorf("service name is empty")
	}
	s.Group = strings.TrimSpace(c.Query("group"))
	s.Version = strings.TrimSpace(c.Query("version"))
	s.Mesh = strings.TrimSpace(c.Query("mesh"))
	return nil
}

func (s *BaseServiceReq) ServiceKey() string {
	return s.ServiceName + constants.ColonSeparator + s.Version + constants.ColonSeparator + s.Group
}

type ServiceMethodDetailReq struct {
	BaseServiceReq

	MethodName string `form:"methodName" json:"methodName"`
	Signature  string `form:"signature" json:"signature"`
}

func (s *ServiceMethodDetailReq) Query(c *gin.Context) error {
	if err := s.BaseServiceReq.Query(c); err != nil {
		return err
	}
	s.MethodName = strings.TrimSpace(c.Query("methodName"))
	if strutil.IsBlank(s.MethodName) {
		return fmt.Errorf("method name is empty")
	}
	s.Signature = strings.TrimSpace(c.Query("signature"))
	if strutil.IsBlank(s.Signature) {
		return fmt.Errorf("signature is empty")
	}
	return nil
}

type ServiceMethodSummaryResp struct {
	MethodName     string   `json:"methodName"`
	ParameterTypes []string `json:"parameterTypes"`
	Signature      string   `json:"signature,omitempty"`
}

type ServiceMethodParameter struct {
	Name string `json:"name"`
	Type string `json:"type"`
}

type ServiceMethodDetailResp struct {
	MethodName     string                   `json:"methodName"`
	Signature      string                   `json:"signature,omitempty"`
	ParameterTypes []string                 `json:"parameterTypes"`
	Parameters     []ServiceMethodParameter `json:"parameters"`
	ReturnType     string                   `json:"returnType"`
	Types          []ServiceMethodTypeResp  `json:"types"`
}

type ServiceMethodTypeResp struct {
	Type       string            `json:"type"`
	Properties map[string]string `json:"properties"`
	Items      []string          `json:"items"`
	Enums      []string          `json:"enums"`
}

const DefaultServiceGenericInvokeTimeoutMs int64 = 3000

type ServiceGenericInvokeReq struct {
	BaseServiceReq

	InstanceName string            `json:"instanceName"`
	MethodName   string            `json:"methodName"`
	Signature    string            `json:"signature"`
	Args         []json.RawMessage `json:"args"`
	TimeoutMs    int64             `json:"timeoutMs"`
	Attachments  map[string]string `json:"attachments"`
}

func (s *ServiceGenericInvokeReq) Validate() error {
	s.Mesh = strings.TrimSpace(s.Mesh)
	if strutil.IsBlank(s.Mesh) {
		return fmt.Errorf("mesh is empty")
	}

	s.ServiceName = strings.TrimSpace(s.ServiceName)
	if strutil.IsBlank(s.ServiceName) {
		return fmt.Errorf("service name is empty")
	}

	s.MethodName = strings.TrimSpace(s.MethodName)
	if strutil.IsBlank(s.MethodName) {
		return fmt.Errorf("method name is empty")
	}

	s.Signature = strings.TrimSpace(s.Signature)
	if strutil.IsBlank(s.Signature) {
		return fmt.Errorf("signature is empty")
	}

	s.InstanceName = strings.TrimSpace(s.InstanceName)
	if strutil.IsBlank(s.InstanceName) {
		return fmt.Errorf("instance name is empty")
	}

	s.Group = strings.TrimSpace(s.Group)
	s.Version = strings.TrimSpace(s.Version)

	if s.TimeoutMs <= 0 {
		s.TimeoutMs = DefaultServiceGenericInvokeTimeoutMs
	}

	return nil
}

type ServiceGenericInvokeResp struct {
	ElapsedMs int64 `json:"elapsedMs"`
	RawResult any   `json:"rawResult"`
}

File: ui-vue3/src/mocks/handlers/service.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { http, type HttpHandler } from 'msw'
import { success, base } from '../utils'
import type {
  ServiceSearchItem,
  ServiceMethod,
  ServiceMethodDetail,
  ProviderInstance,
  ArgumentRouteConfig,
  PaginatedData
} from '@/types/api'

const serviceList: ServiceSearchItem[] = [
  {
    serviceName: 'org.apache.dubbo.samples.UserService',
    versionGroups: [
      { version: '1.0.0', group: 'group1' },
      { version: '1.0.0', group: null },
      { version: null, group: 'group1' },
      { version: null, group: null }
    ],
    avgQPS: 6,
    avgRT: '194ms',
    requestTotal: 200
  },
  {
    serviceName: 'org.apache.dubbo.samples.OrderService',
    versionGroups: [
      { version: '1.0.0', group: 'group1' },
      { version: '1.0.0', group: null },
      { version: null, group: 'group1' },
      { version: null, group: null }
    ],
    avgQPS: 13,
    avgRT: '189ms',
    requestTotal: 164
  },
  {
    serviceName: 'org.apache.dubbo.samples.DetailService',
    versionGroups: [
      { version: '1.0.0', group: 'group1' },
      { version: '1.0.0', group: null },
      { version: null, group: 'group1' },
      { version: null, group: null }
    ],
    avgQPS: 0.5,
    avgRT: '268ms',
    requestTotal: 1324
  },
  {
    serviceName: 'org.apache.dubbo.samples.PayService',
    versionGroups: [
      { version: '1.0.0', group: 'group1' },
      { version: '1.0.0', group: null },
      { version: null, group: 'group1' },
      { version: null, group: null }
    ],
    avgQPS: 9,
    avgRT: '346ms',
    requestTotal: 189
  },
  {
    serviceName: 'org.apache.dubbo.samples.CommentService',
    versionGroups: [
      { version: '1.0.0', group: 'group1' },
      { version: '1.0.0', group: null },
      { version: null, group: 'group1' },
      { version: null, group: null }
    ],
    avgQPS: 8,
    avgRT: '936ms',
    requestTotal: 200
  },
  {
    serviceName: 'org.apache.dubbo.samples.RepayService',
    versionGroups: [
      { version: '1.0.0', group: 'group1' },
      { version: '1.0.0', group: null },
      { version: null, group: 'group1' },
      { version: null, group: null }
    ],
    avgQPS: 17,
    avgRT: '240ms',
    requestTotal: 146
  },
  {
    serviceName: 'org.apche.dubbo.samples.TransportService',
    versionGroups: [
      { version: '1.0.0', group: 'group1' },
      { version: '1.0.0', group: null },
      { version: null, group: 'group1' },
      { version: null, group: null }
    ],
    avgQPS: 43,
    avgRT: '89ms',
    requestTotal: 367
  },
  {
    serviceName: 'org.apche.dubbo.samples.DistributionService',
    versionGroups: [
      { version: '1.0.0', group: 'group1' },
      { version: '1.0.0', group: null },
      { version: null, group: 'group1' },
      { version: null, group: null }
    ],
    avgQPS: 4,
    avgRT: '78ms',
    requestTotal: 145
  }
]

const serviceMethods: ServiceMethod[] = [
  {
    methodName: 'query',
    parameterTypes: ['java.lang.String'],
    signature: 'java.lang.String->org.apache.demo.UserDTO'
  },
  {
    methodName: 'create',
    parameterTypes: ['org.apache.demo.UserCreateReq'],
    signature: 'org.apache.demo.UserCreateReq->java.lang.Boolean'
  },
  {
    methodName: 'delete',
    parameterTypes: ['java.lang.String'],
    signature: 'java.lang.String->java.lang.Boolean'
  }
]

const methodDetails: Record<string, ServiceMethodDetail> = {
  query: {
    methodName: 'query',
    signature: 'java.lang.String->org.apache.demo.UserDTO',
    parameterTypes: ['java.lang.String'],
    parameters: [{ name: 'id', type: 'java.lang.String' }],
    returnType: 'org.apache.demo.UserDTO',
    types: [
      {
        type: 'org.apache.demo.UserDTO',
        properties: { id: 'java.lang.String', name: 'java.lang.String', age: 'java.lang.Integer' },
        items: [],
        enums: []
      }
    ]
  },
  create: {
    methodName: 'create',
    signature: 'org.apache.demo.UserCreateReq->java.lang.Boolean',
    parameterTypes: ['org.apache.demo.UserCreateReq'],
    parameters: [{ name: 'req', type: 'org.apache.demo.UserCreateReq' }],
    returnType: 'java.lang.Boolean',
    types: [
      {
        type: 'org.apache.demo.UserCreateReq',
        properties: {
          name: 'java.lang.String',
          age: 'java.lang.Integer',
          email: 'java.lang.String'
        },
        items: [],
        enums: []
      }
    ]
  },
  delete: {
    methodName: 'delete',
    signature: 'java.lang.String->java.lang.Boolean',
    parameterTypes: ['java.lang.String'],
    parameters: [{ name: 'id', type: 'java.lang.String' }],
    returnType: 'java.lang.Boolean',
    types: []
  }
}

const providerInstances: ProviderInstance[] = [
  { name: 'dubbo-provider-0', appName: 'dubbo-sample-provider', ip: '10.20.30.11' },
  { name: 'dubbo-provider-1', appName: 'dubbo-sample-provider', ip: '10.20.30.12' }
]

const argumentRouteConfig: ArgumentRouteConfig = {
  args: [
    {
      type: 'header',
      key: 'X-User-Id',
      operator: '=',
      value: '123',
      serviceName: 'org.apache.dubbo.samples.UserService'
    }
  ]
}

export const serviceHandlers: HttpHandler[] = [
  http.get(`${base}/service/search`, () =>
    success<PaginatedData<ServiceSearchItem>>({
      pageInfo: { Total: 8, NextOffset: '0' },
      list: serviceList
    })
  ),

  http.get(`${base}/service/metric-dashboard`, () =>
    success(
      'http://8.147.104.101:3000/d/a0b114ca-edf7-4dfe-ac2c-34a4fc545fed/service?orgId=1&refresh=1m'
    )
  ),

  http.get(`${base}/service/trace-dashboard`, () =>
    success(
      'http://8.147.104.101:3000/d/e968a89b-f03d-42e3-8ad3-930ae815cb0f/service?orgId=1&refresh=1m'
    )
  ),

  http.get(`${base}/service/config/timeout`, () => success({ timeout: 3000 })),

  http.put(`${base}/service/config/timeout`, () => success(null)),

  http.get(`${base}/service/config/retry`, () => success({ retry: 3 })),

  http.put(`${base}/service/config/retry`, () => success(null)),

  http.get(`${base}/service/config/regionPriority`, () => success({ enable: true })),

  http.put(`${base}/service/config/regionPriority`, () => success(null)),

  http.get(`${base}/service/config/argumentRoute`, () => success(argumentRouteConfig)),

  http.put(`${base}/service/config/argumentRoute`, () => success(null)),

  http.get(`${base}/service/provider-instances`, () => success(providerInstances)),

  http.get(`${base}/service/methods`, () => success(serviceMethods)),

  http.get(`${base}/service/method/detail`, ({ request }) => {
    const url = new URL(request.url)
    const methodName = url.searchParams.get('methodName') || 'query'
    return success(methodDetails[methodName] || methodDetails.query)
  }),

  http.post(`${base}/service/generic/invoke`, () =>
    success({ elapsedMs: 12, rawResult: { id: '1001', name: 'Alice', age: 18 } })
  )
]
